syntax = "proto3";

package kinfer;

/* ------ *
 * Tensor *
 * ------ */

message TensorValue {
    repeated int32 shape = 2;
    repeated float data = 3;
}

/* ---------------- *
 * Joint Positions *
 * ---------------- */

message JointPosition {
    string joint_name = 1;
    float degrees = 2;
}

message JointPositionsValue {
    repeated JointPosition positions = 1;
}

/* ---------------- *
 * Joint Velocities *
 * ---------------- */

message JointVelocity {
    string joint_name = 1;
    float degrees_per_second = 2;
}

message JointVelocitiesValue {
    repeated JointVelocity velocities = 1;
}

/* -------------- *
 * Joint Torques *
 * -------------- */

message JointTorque {
    string joint_name = 1;
    float newton_meters = 2;
}

message JointTorquesValue {
    repeated JointTorque torques = 1;
}

/* ------------ *
 * Camera Frame *
 * ------------ */

message CameraFrameValue {
    int32 width = 1;
    int32 height = 2;
    int32 channels = 3;
    repeated float data = 4;
}

/* ----------- *
 * Audio Frame *
 * ----------- */

message AudioFrameValue {
    int32 channels = 1;
    int32 sample_rate = 2;
    repeated float data = 3;
}

/* --- *
 * IMU *
 * --- */

message IMUValue {
    float x = 1;
    float y = 2;
    float z = 3;
    float w = 4;
}

/* --------- *
 * Timestamp *
 * --------- */

message TimestampValue {
    int64 seconds = 1;
    int32 nanos = 2;
    int64 start_seconds = 3;
    int32 start_nanos = 4;
}

/* ----- *
 * Value *
 * ----- */

message Value {
    string value_name = 1;
    oneof value {
        TensorValue tensor = 2;
        JointPositionsValue joint_positions = 3;
        JointVelocitiesValue joint_velocities = 4;
        JointTorquesValue joint_torques = 5;
        CameraFrameValue camera_frame = 6;
        AudioFrameValue audio_frame = 7;
        IMUValue imu = 8;
        TimestampValue timestamp = 9;
    }
}

enum ValueType {
    TENSOR = 0;
    JOINT_POSITIONS = 1;
    JOINT_VELOCITIES = 2;
    JOINT_TORQUES = 3;
    CAMERA_FRAME = 4;
    AUDIO_FRAME = 5;
    IMU = 6;
    TIMESTAMP = 7;
}

message ValueSchema {
    string name = 1;
    ValueType value_type = 2;
}

/* ------ *
 * Inputs *
 * ------ */

message Input {
    repeated Value inputs = 1;
}

message InputSchema {
    repeated ValueSchema inputs = 1;
}

/* ------- *
 * Outputs *
 * ------- */

message Output {
    repeated Value outputs = 1;
}

message OutputSchema {
    repeated ValueSchema outputs = 1;
}
